### PR TITLE
Add initial CI pipeline with UI tests and axe scanning

### LINKED ISSUE #1024: Create CI pipeline
### Issue type

enhancement

### Which version of the app?

WinUI 3 Gallery

### Description

Use this pipeline to do test automation, performance testing, and accessibility testing. 

Start with simple navigation test in this pipeline.

### Screenshots

_No response_

### Windows version

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by chingucoding**
This PR adds the initial CI pipeline with azure devops, re-enable the UI tests and add Axe Accessibility scanning.

Closes #1024

### CODE REVIEW

**Comment by kmahone** [comment]
We should avoid adding a .pfx key to the repo. This can cause issues, since they expire and also get flagged by scanning tools as being 'secrets' even in cases such as this where it is just for testing.

In WinUI, we use a script to generate the .pfx key at build time  [GenerateTestPfx.ps1](https://github.com/microsoft/microsoft-ui-xaml/blob/main/GenerateTestPfx.ps1)
A similar approach might work well here.

File: WinUIGallery/Package.WAP.Dev.appxmanifest
Comment (kmahone) [inline WinUIGallery/Package.WAP.Dev.appxmanifest:10]: >Publisher="CN=Microsoft Corporation" [](http://example.com/codeflow?start=57&length=36)

If this is being used for testing purposes, we could update the Publisher to something like "CN=WinUI-Gallery-Test" so it is clear that it is not an official build of something from Microsoft.

File: UITests/TestBase.cs
Comment (kmahone) [inline UITests/TestBase.cs:34]: Can we do something better than thread sleeps here? This seems unreliable.

E.g. can we wait for a certain piece of UI to update instead?

**Comment by kmahone** [comment]
> global using OpenQA.Selenium;

I don't see the advantage of doing this. It makes more sense to include the using directives in the files that require them.

---
Refers to: UITests/GlobalUsings.cs:1 in 07c17dd. [](commit_id = 07c17dd72cc19bf9d05796e75f5d9e6a14e767b9, deletion_comment = False)

File: UITests/TestBase.cs
Comment (kmahone) [inline UITests/TestBase.cs:35]: We should add explicit Axe test methods, instead of having this validation piggy-back on other tests via this helper method.

Comment (kmahone) [reply UITests/TestBase.cs:35]: E.g. Tests\Button.cs should be updated to have
```
[TestMethod]
public void ValidateAccessibilityWithAxe()
```
or something like that.

File: UITests/TestBase.cs
Comment (kmahone) [inline UITests/TestBase.cs:46]: What is the purpose of this Cleanup step?

**Review by kmahone — CHANGES_REQUESTED** [review]
:clock1:

File: UITests/TestBase.cs
Comment (chingucoding) [reply UITests/TestBase.cs:36]: I'm not really sure how you would do that with selenium, if you have any pointers I would love to do that.

**Comment by chingucoding** [comment]
I think I addressed most of your feedback @kmahone, the global usings file is gone. accessibility testing is an explicit test now. Regarding waiting for an element, I would appreciate some pointers on how to set this up.

**Comment by kmahone** [comment]
> I think I addressed most of your feedback @kmahone, the global usings file is gone. accessibility testing is an explicit test now. Regarding waiting for an element, I would appreciate some pointers on how to set this up.

Thanks. I've updated the comment with a suggested approach for this.

File: UITests/TestBase.cs
Comment (chingucoding) [reply UITests/TestBase.cs:36]: I also thought about that API, but its quite sophisticated. IIRC, part of that infra was also moved into a package right? Or was that something else.

Regarding the polling approach, I'm not sure if there is something sensible we want to wait for. The idea was to wait for the app to navigate to the page but I don't think we can "poll" for this without introducing UIA elements.

Comment (kmahone) [reply UITests/TestBase.cs:36]: In the case of a page navigate you could wait until a known element on the page is available in the UIA tree.

File: UITests/TestBase.cs
Comment (kmahone) [inline UITests/TestBase.cs:64]: >} [](http://example.com/codeflow?start=3&length=1)

We should throw in the case where we never find the element.

**Review by kmahone — APPROVED** [review]

**Comment by kmahone** [comment]
Looks good to me. @karkarl should take a look and approve also before merge.

File: UITests/TestBase.cs
Comment (chingucoding) [reply UITests/TestBase.cs:63]: Gread idea, fixed that.

**Comment by chingucoding** [comment]
After/before merging, of course some Azure DevOps project also needs to be configured which is not something that I can do.

**Comment by kmahone** [comment]
> After/before merging, of course some Azure DevOps project also needs to be configured which is not something that I can do.

Yes. Karen is going to set that up.

**Review by karkarl — APPROVED** [review]

**Pull request merged**
